### DISCUSSION TITLE
Combination of layout 2 and space (tap) is not working

### DISCUSSION THREAD

**Discussion by KoichaDev** [Yowkees/keyball#445] [Q&A] [answered]
Hello everyone! 

I have a combination that I wish to work on my Keyball39. 

If I tap the `KC_SPACE` then it will always act as space. If I hold down the `KC_SPACE`, then it will switch to layout 2 as momentarily. If let go the `KC_SPACE`, then the layout 2 will be toggles off automatically, and the `KC_SPACE` will be back as the default behavior. 

This behavior is not working as intended if I by mistake double hitting the space key and holding the space key down, and it will not switch to layout 2, instead it acts as default space.

![SCR-20231211-1zl.png](https://github.com/Yowkees/keyball/assets/22475147/4eb34289-10b9-45ac-95db-20edf1b029ba) 

I have attached a screen shot as how I would the key combination to work. 

I would love to hear if someone else know how to resolve this. 🙏

**Comment by Yowkees** [comment]
Hi. Thank you very much for choosing Keyball39:-)

Can I see the keymap of Layer2? I think there is problem which key is assigned on this key in Layer2.
Thank you.

**Reply by KoichaDev** [reply]
![SCR-20231211-5jd](https://github.com/Yowkees/keyball/assets/22475147/94fe52fb-4d88-4055-b06b-acf6eac8c540)

Sure! Here is the image for the layout 2 above  

![SCR-20231211-5jp](https://github.com/Yowkees/keyball/assets/22475147/26ab6bde-5ddc-4378-89ff-dd39e2e72337)

I have attached the image above for layout 1 as well in case you need to see it

**Reply by KoichaDev** [reply]
![SCR-20231211-f6g](https://github.com/Yowkees/keyball/assets/22475147/519cba11-d176-41fc-a6cb-de4f3f4021c5)

I notice on the screenshot above that the left keyboard is working fine as the behavior I want, but not on the right side. Any idea why this is happening?

**Reply by KoichaDev** [reply]
@Yowkees any idea if this is maybe a firmware bug that needs to be updated or something?

**Reply by KoichaDev** [reply]
@Yowkees 

I found a solution to fix this, which is called `quick tap term`, but it requires to modify the firmware to make it work:

https://docs.qmk.fm/#/tap_hold?id=quick-tap-term

How am I supposed to do this? I am hitting the space button by accident multiple times, and it drives me crazy, because I type fast.

**Comment by Yowkees** [comment]
Hi.

"Space" key has has two functions for both tap and hold.
Thats why coexistence of space and layer(2) on one key is difficult to use for different individuals.

To make your preferred behavior depending on the time to hold, please edit and build the source code of Keyball. It's open source.
I think you can also use the quick tap term you mentioned.

Thanks.

**Reply by KoichaDev** [reply]
Are there any guides how to build/modify the existing source code step-by-step?

 I am only familiar using the remap web application to change existing key press only

**Reply by KoichaDev** [reply, marked as answer]
Great news! I found out how to resolve this! It took me some time to figure this out! 

In case anybody reads this and wishes to either have the same functionality as mine or need to change something on the existing firmware: 

1. [Follow this guide here and go step 1, 2 and 3](https://docs.qmk.fm/#/newbs_getting_started)
2. [Follow this next step here](https://github.com/Yowkees/keyball/blob/main/qmk_firmware/keyboards/keyball/readme.md), but stop on __step 4__ of the guide.
3. Open your editor and locate the `qmk` folder on your local computer, and navigate to `keyboards/keyball/keyball39/keymaps/via/config.h` and then paste `#define QUICK_TAP_TERM 0` end of the code line and save the file.
4. Then go back to the [step 4](https://github.com/Yowkees/keyball/blob/main/qmk_firmware/keyboards/keyball/readme.md) from the guide and then enter `make SKIP_GIT=yes keyball/keyball39:via` on your terminal and run it!
5. When the build process is finished, then [go to this Remap catalogue webite](https://remap-keys.app/catalog/k895xiCpsM5zlYZCPTEN/firmware) and [click on the logo](https://github.com/Yowkees/keyball/assets/22475147/a0ce09d5-2756-4ac8-b4ca-94ed4251e1fe) on the top right corner.
6. It will open a modal to ask you to drop your firmware on the page. This is where you built your custom firmware on step 4 above. The file should be named `keyball_keyball39_via.hex`. Take that file and drop it on the remap page.
7. Click on `flash`, and then double press on your keyboard the reset button, then it should give you a list where you can choose your pro micro. That is the one you want to flash it on your keyboard.
8. When flashed, then you have your modified firmware version of your firmware!

Happy modding your keyboard! I know that I am pleased with the result of my keyboard right now! 🎉
